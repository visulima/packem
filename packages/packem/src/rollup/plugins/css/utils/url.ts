import { isAbsolute, normalize, parse } from "@visulima/path";
import { isRelative } from "@visulima/path/utils";

/**
 * Checks if the URL starts with a tilde followed by a digit, '@', or a letter (case-insensitive)
 * @param url
 * @returns
 */
export const hasModuleSpecifier = (url: string): boolean => /^~[\d@A-Z]/i.test(url);

// handle importing Sass partials in node_modules
// @import ~foo/bar/partial where "partial" has the filename "_partial.scss".
export const getUrlOfPartial = (url: string): string => {
    const { base, dir } = parse(url);

    return dir ? `${normalize(dir)}/_${base}` : `_${base}`;
};

export const normalizeUrl = (url: string): string => {
    if (hasModuleSpecifier(url)) {
        return normalize(url.slice(1));
    }

    if (isAbsolute(url) || isRelative(url)) {
        return normalize(url);
    }

    return `./${normalize(url)}`;
};
