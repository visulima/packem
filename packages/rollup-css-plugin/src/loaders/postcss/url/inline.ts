import liteMime from "mime/lite";

export default (file: string, source: Uint8Array): string => {
    const mime = liteMime.getType(file) ?? "application/octet-stream";
    const data = Buffer.from(source).toString("base64");

    return `data:${mime};base64,${data}`;
};
