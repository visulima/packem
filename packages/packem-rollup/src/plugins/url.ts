/**
 * Modified copy of https://github.com/rollup/plugins/blob/master/packages/url/src/index.js
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 RollupJS Plugin Contributors (https://github.com/rollup/plugins/graphs/contributors)
 */
import crypto from "node:crypto";
import { createReadStream, createWriteStream } from "node:fs";
import { readFile, stat } from "node:fs/promises";

import type { FilterPattern } from "@rollup/pluginutils";
import { createFilter } from "@rollup/pluginutils";
import { ensureDir } from "@visulima/fs";
import { basename, dirname, extname, join, relative } from "@visulima/path";
import mime from "mime";
import type { Plugin } from "rollup";

import { svgEncoder } from "../utils";

const copy = async (source: string, destination: string): Promise<void> => {
    await new Promise((resolve, reject) => {
        const read = createReadStream(source);

        read.on("error", reject);

        const write = createWriteStream(destination);

        write.on("error", reject);
        write.on("finish", () => resolve(undefined));

        read.pipe(write);
    });
};

export interface UrlOptions {
    /**
     * The destination dir to copy assets, usually used to rebase the assets according to HTML files.
     * @type {string}
     */
    destDir?: string;
    /**
     * If false, will prevent files being emitted by this plugin. This is useful for when you are using Rollup to emit both a client-side and server-side bundle.
     * @type {boolean}
     * @default true
     */
    emitFiles: boolean;
    /**
     * A picomatch pattern, or array of patterns, which specifies the files in the build the plugin
     * should _ignore_.
     *
     * By default, no files are ignored.
     * @type {FilterPattern}
     */
    exclude?: FilterPattern;
    /**
     * If emitFiles is true, this option can be used to rename the emitted files. It accepts the following string replacements:
     * [hash] - The hash value of the file's contents
     * [name] - The name of the imported file (without its file extension)
     * [extname] - The extension of the imported file (including the leading .)
     * [dirname] - The parent directory name of the imported file (including trailing /)
     * @type {string}
     * @default [hash][extname]
     */
    fileName: string;
    /**
     * A picomatch pattern, or array of patterns, which specifies the files in the build the plugin
     * should operate on.
     * By default, the png,jpg,jpeg,gif,svg,webp files are targeted.
     * @type {FilterPattern}
     */
    include?: FilterPattern;
    /**
     * The file size limit for inline files.
     * If a file exceeds this limit, it will be copied to the destination folder and the hashed filename will be provided instead.
     * If limit is set to 0 all files will be copied.
     * @type {number}
     * @default 14336 (14kb)
     */
    limit: number;
    /**
     * A string which will be added in front of filenames when they are not inlined but are copied.
     * @type {string}
     */
    publicPath?: string;

    /**
     * When using the [dirname] replacement in fileName, use this directory as the source directory from which to create the file path rather than the parent directory of the imported file. For example:
     * @example
     * ```js
     *    src/path/to/file.js
     *
     *    import png from './image.png';
     *    rollup.config.js
     *
     *    url({
     *      fileName: '[dirname][hash][extname]',
     *      sourceDir: path.join(__dirname, 'src')
     *    });
     *
     *    Emitted File: path/to/image.png
     * ```
     */
    sourceDir?: string;
}

export const urlPlugin = ({
    destDir: destinationDirectory,
    emitFiles,
    exclude,
    fileName,
    include,
    limit,
    publicPath,
    sourceDir: sourceDirectory,
}: UrlOptions): Plugin => {
    const filter = createFilter(include, exclude);
    const copies: Record<string, string> = {};

    return <Plugin>{
        async generateBundle(outputOptions: { dir?: string; file?: string }) {
            if (!emitFiles) {
                return;
            }

            const base = destinationDirectory ?? outputOptions.dir ?? dirname(outputOptions.file ?? "");

            await ensureDir(base);

            await Promise.all(
                Object.keys(copies).map(async (name) => {
                    const output = copies[name] as string;
                    const outputDirectory = join(base, dirname(output));

                    await ensureDir(outputDirectory);
                    await copy(name, join(base, output));
                }),
            );
        },

        async load(id: string) {
            if (!filter(id)) {
                return undefined;
            }

            this.addWatchFile(id);

            const [stats, buffer] = await Promise.all([stat(id), readFile(id)]);

            let data: string;

            if ((limit && stats.size > limit) || limit === 0) {
                const hash = crypto.createHash("sha1").update(buffer).digest("hex").slice(0, 16);
                const extension = extname(id);
                const name = basename(id, extension);
                const relativeDirectory = sourceDirectory ? relative(sourceDirectory, dirname(id)) : basename(dirname(id));

                const outputFileName = fileName
                    .replaceAll("[hash]", hash)
                    .replaceAll("[extname]", extension)
                    .replaceAll("[dirname]", relativeDirectory === "" ? "" : `${relativeDirectory}/`)
                    .replaceAll("[name]", name);

                data = join(publicPath ?? "", outputFileName);

                copies[id] = outputFileName;
            } else {
                const mimetype = mime.getType(id);

                if (mimetype === undefined) {
                    throw new Error(`Could not determine mimetype for ${id}`);
                }

                const isSVG = mimetype === "image/svg+xml";

                data = isSVG ? svgEncoder(buffer) : buffer.toString("base64");

                data = `data:${mimetype};base64,${data}`;
            }

            return `export default "${data}"`;
        },
        name: "packem:url",
    };
};
