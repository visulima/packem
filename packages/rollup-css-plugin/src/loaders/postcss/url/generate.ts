import { getHash } from "@visulima/packem-share/utils";
import { basename, parse } from "@visulima/path";

import { HASH_REGEXP } from "../constants";

export default (placeholder: string, file: string, source: Uint8Array): string => {
    const { base, dir, ext, name } = parse(file);
    const hash = getHash(`${base}:${Buffer.from(source).toString()}`);
    const match = HASH_REGEXP.exec(placeholder);
    const hashLength = match && Number.parseInt(match[1] as string, 10);

    return placeholder
        .replace("[dir]", basename(dir))
        .replace("[name]", name)
        .replace("[extname]", ext)
        .replace(".[ext]", ext)
        .replace("[ext]", ext.slice(1))
        .replace(HASH_REGEXP, hashLength ? hash.slice(0, hashLength) : hash.slice(0, 8));
};
