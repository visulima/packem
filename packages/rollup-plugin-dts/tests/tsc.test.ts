import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { rolldownBuild } from "@sxzz/test-utils";
import { glob } from "tinyglobby";
import { describe, expect } from "vitest";

import { dts } from "../src/index.js";

const dirname = path.dirname(fileURLToPath(import.meta.url));

describe("tsc", () => {
    it("typescript compiler", async () => {
        const root = path.resolve(dirname, "fixtures/tsc");
        const { snapshot } = await rolldownBuild(
            [path.resolve(root, "entry1.ts"), path.resolve(root, "entry2.ts")],
            [
                dts({
                    compilerOptions: {
                        isolatedDeclarations: false,
                        skipLibCheck: true,
                    },
                    emitDtsOnly: true,
                    isolatedDeclarations: false,
                }),
            ],
        );

        expect(snapshot.replaceAll(/\/\/#region.*/g, "")).toMatchSnapshot();
    });

    it("multi declarations", async () => {
        const { snapshot } = await rolldownBuild(
            path.resolve(dirname, "fixtures/multi-decls/index.ts"),
            [
                dts({
                    compilerOptions: { isolatedDeclarations: false },
                    emitDtsOnly: true,
                }),
            ],
        );

        expect(snapshot).toMatchSnapshot();
    });

    it("composite references", async () => {
        const root = path.resolve(dirname, "fixtures/composite-refs");

        // The outDir in tsconfig files.
        const tempDir = path.resolve(root, "temp");

        // Ensure .tsbuildinfo files do not exist before the test
        await fs.rm(tempDir, { force: true, recursive: true });

        const { snapshot } = await rolldownBuild(
            [
                path.resolve(root, "dir1/input1.ts"),
                path.resolve(root, "dir2/input2.ts"),
            ],
            [
                dts({
                    compilerOptions: { isolatedDeclarations: false },
                    tsconfig: path.resolve(root, "tsconfig.json"),
                }),
            ],
        );

        expect(snapshot).toMatchSnapshot();

        // Ensure .tsbuildinfo files are not created after the test
        const tsBuildInfoFiles = await glob("**/*.tsbuildinfo", {
            absolute: false,
            cwd: tempDir,
        });

        expect(tsBuildInfoFiles).toHaveLength(0);
    });

    it("composite references incremental", async () => {
        const root = path.resolve(dirname, "fixtures/composite-refs-incremental");

        // The outDir in tsconfig files.
        const tempDir = path.resolve(root, "temp");

        // Ensure .tsbuildinfo files do not exist before the test
        await fs.rm(tempDir, { force: true, recursive: true });

        const { snapshot } = await rolldownBuild(
            [
                path.resolve(root, "dir1/input1.ts"),
                path.resolve(root, "dir2/input2.ts"),
            ],
            [
                dts({
                    compilerOptions: { isolatedDeclarations: false },
                    tsconfig: path.resolve(root, "tsconfig.json"),
                }),
            ],
        );

        expect(snapshot).toMatchSnapshot();

        // Ensure .tsbuildinfo files are created after the test
        const tsBuildInfoFiles = await glob("**/*.tsbuildinfo", {
            absolute: false,
            cwd: tempDir,
        });

        expect(tsBuildInfoFiles.sort()).toMatchInlineSnapshot(`
            [
              "dir1/tsconfig.1.tsbuildinfo",
              "dir2/tsconfig.2.tsbuildinfo",
            ]
        `);
    });

    it("vue-sfc w/ ts-compiler", async () => {
        const root = path.resolve(dirname, "fixtures/vue-sfc");
        const { snapshot } = await rolldownBuild(path.resolve(root, "main.ts"), [
            dts({
                emitDtsOnly: true,
                vue: true,
            }),
        ]);

        expect(snapshot).toMatchSnapshot();
    });

    it("jsdoc", async () => {
        const { snapshot } = await rolldownBuild(
            path.resolve(dirname, "fixtures/jsdoc.ts"),
            [dts({ isolatedDeclarations: false })],
        );

        expect(snapshot).toMatchSnapshot();
    });
});
