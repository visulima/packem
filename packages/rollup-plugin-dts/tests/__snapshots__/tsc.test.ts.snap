// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tsc > composite references 1`] = `
"// input1.d.ts
//#region tests/fixtures/composite-refs/dir1/input1.d.ts
declare const input1 = "input1";
//#endregion
export { input1 };
// input1.js
//#region tests/fixtures/composite-refs/dir1/input1.ts
const input1 = "input1";

//#endregion
export { input1 };
// input2.d.ts
//#region tests/fixtures/composite-refs/dir2/input2.d.ts
declare const input2 = "input2";
//#endregion
export { input2 };
// input2.js
//#region tests/fixtures/composite-refs/dir2/input2.ts
const input2 = "input2";

//#endregion
export { input2 };"
`;

exports[`tsc > composite references incremental 1`] = `
"// input1.d.ts
//#region tests/fixtures/composite-refs-incremental/dir1/input1.d.ts
declare const input1 = "input1";
//#endregion
export { input1 };
// input1.js
//#region tests/fixtures/composite-refs-incremental/dir1/input1.ts
const input1 = "input1";

//#endregion
export { input1 };
// input2.d.ts
//#region tests/fixtures/composite-refs-incremental/dir2/input2.d.ts
declare const input2 = "input2";
//#endregion
export { input2 };
// input2.js
//#region tests/fixtures/composite-refs-incremental/dir2/input2.ts
const input2 = "input2";

//#endregion
export { input2 };"
`;

exports[`tsc > jsdoc 1`] = `
"// jsdoc.d.ts
import { Plugin } from "rolldown";

//#region tests/fixtures/jsdoc.d.ts

/**
 * with imports
 */
interface RollupMdiFontminOptions extends Plugin {}
/**
 * named export
 */
declare function fn(): number;
/**
 * options
 */
interface Options {
  /**
   * interface member
   */
  foo: string;
}
/**
 * type alias
 */
type Foo = string;
/**
 * default export
 */
declare const _default: number;
//#endregion
export { Foo, Options, RollupMdiFontminOptions, _default as default, fn };
// jsdoc.js
//#region tests/fixtures/jsdoc.ts
/**
* named export
*/
function fn() {
	return 42;
}
/**
* default export
*/
var jsdoc_default = fn();

//#endregion
export { jsdoc_default as default, fn };"
`;

exports[`tsc > jsdoc in js 1`] = `
"// main.d.ts
//#region tests/fixtures/jsdoc-js/main.d.ts
/**
 * Does bar.
 *
 * @param {string} baz
 * @returns {void}
 */
declare function foo(baz: string): void;
//#endregion
export { foo };"
`;

exports[`tsc > multi declarations 1`] = `
"// index.d.ts
//#region tests/fixtures/multi-decls/mod.d.ts
declare const a: 1, b: 2;
//#endregion
//#region tests/fixtures/multi-decls/index.d.ts
declare const a$1 = 3,
  b$1 = 3;
//#endregion
export { a, b, a$1 as c, b$1 as d };"
`;

exports[`tsc > references 1`] = `
"// index.d.ts
//#region tests/fixtures/refs/src/index.d.ts
declare const a = 1;
//#endregion
export { a };
// index.js
//#region tests/fixtures/refs/src/index.ts
const a = 1;

//#endregion
export { a };"
`;

exports[`tsc > typescript compiler 1`] = `
"// entry1.d.ts

declare const str = "foo";
declare const num = 42;
declare const bool = true;
declare function fn(arg: typeof str): "foo" | 1;
//#endregion
export { bool, fn, num, str };
// entry2.d.ts

interface Unused {
  foo: string;
}
//#endregion
export { Unused };"
`;

exports[`tsc > vue-sfc w/ ts-compiler 1`] = `
"// main.d.ts
import * as vue0 from "vue";

//#region tests/fixtures/vue-sfc/App.vue.d.ts
type __VLS_Props = {
  foo: string;
};
declare global {
  interface Window {
    foo: string;
  }
}
declare const _default: vue0.DefineComponent<__VLS_Props, {}, {}, {}, {}, vue0.ComponentOptionsMixin, vue0.ComponentOptionsMixin, {}, string, vue0.PublicProps, Readonly<__VLS_Props> & Readonly<{}>, {}, {}, {}, {}, string, vue0.ComponentProvideOptions, false, {}, any>;
//#endregion
export { _default as App };"
`;

exports[`tsc > vue-sfc w/ ts-compiler w/ vueCompilerOptions in tsconfig 1`] = `
"// main.d.ts
import * as vue0 from "vue";

//#region tests/fixtures/vue-sfc-fallthrough/App.vue.d.ts
declare const _default: vue0.DefineComponent<{
  foo: string;
  bar?: string;
  innerHTML?: string;
  class?: any;
  style?: vue0.StyleValue;
  accesskey?: string;
  contenteditable?: (boolean | "true" | "false") | "inherit" | "plaintext-only";
  contextmenu?: string;
  dir?: string;
  draggable?: boolean | "true" | "false";
  hidden?: (boolean | "true" | "false") | "" | "hidden" | "until-found";
  id?: string;
  inert?: boolean | "true" | "false";
  lang?: string;
  placeholder?: string;
  spellcheck?: boolean | "true" | "false";
  tabindex?: string | number;
  title?: string;
  translate?: "yes" | "no";
  radiogroup?: string;
  role?: string;
  about?: string;
  datatype?: string;
  inlist?: any;
  prefix?: string;
  property?: string;
  resource?: string;
  typeof?: string;
  vocab?: string;
  autocapitalize?: string;
  autocorrect?: string;
  autosave?: string;
  color?: string;
  itemprop?: string;
  itemscope?: boolean | "true" | "false";
  itemtype?: string;
  itemid?: string;
  itemref?: string;
  results?: string | number;
  security?: string;
  unselectable?: "on" | "off";
  inputmode?: "none" | "text" | "tel" | "url" | "email" | "numeric" | "decimal" | "search";
  is?: string;
  'aria-activedescendant'?: string;
  'aria-atomic'?: boolean | "true" | "false";
  'aria-autocomplete'?: "none" | "inline" | "list" | "both";
  'aria-busy'?: boolean | "true" | "false";
  'aria-checked'?: (boolean | "true" | "false") | "mixed";
  'aria-colcount'?: string | number;
  'aria-colindex'?: string | number;
  'aria-colspan'?: string | number;
  'aria-controls'?: string;
  'aria-current'?: (boolean | "true" | "false") | "page" | "step" | "location" | "date" | "time";
  'aria-describedby'?: string;
  'aria-details'?: string;
  'aria-disabled'?: boolean | "true" | "false";
  'aria-dropeffect'?: "none" | "copy" | "execute" | "link" | "move" | "popup";
  'aria-errormessage'?: string;
  'aria-expanded'?: boolean | "true" | "false";
  'aria-flowto'?: string;
  'aria-grabbed'?: boolean | "true" | "false";
  'aria-haspopup'?: (boolean | "true" | "false") | "menu" | "listbox" | "tree" | "grid" | "dialog";
  'aria-hidden'?: boolean | "true" | "false";
  'aria-invalid'?: (boolean | "true" | "false") | "grammar" | "spelling";
  'aria-keyshortcuts'?: string;
  'aria-label'?: string;
  'aria-labelledby'?: string;
  'aria-level'?: string | number;
  'aria-live'?: "off" | "assertive" | "polite";
  'aria-modal'?: boolean | "true" | "false";
  'aria-multiline'?: boolean | "true" | "false";
  'aria-multiselectable'?: boolean | "true" | "false";
  'aria-orientation'?: "horizontal" | "vertical";
  'aria-owns'?: string;
  'aria-placeholder'?: string;
  'aria-posinset'?: string | number;
  'aria-pressed'?: (boolean | "true" | "false") | "mixed";
  'aria-readonly'?: boolean | "true" | "false";
  'aria-relevant'?: "additions" | "additions removals" | "additions text" | "all" | "removals" | "removals additions" | "removals text" | "text" | "text additions" | "text removals";
  'aria-required'?: boolean | "true" | "false";
  'aria-roledescription'?: string;
  'aria-rowcount'?: string | number;
  'aria-rowindex'?: string | number;
  'aria-rowspan'?: string | number;
  'aria-selected'?: boolean | "true" | "false";
  'aria-setsize'?: string | number;
  'aria-sort'?: "none" | "ascending" | "descending" | "other";
  'aria-valuemax'?: string | number;
  'aria-valuemin'?: string | number;
  'aria-valuenow'?: string | number;
  'aria-valuetext'?: string;
  onCopy?: (payload: ClipboardEvent) => void;
  onCut?: (payload: ClipboardEvent) => void;
  onPaste?: (payload: ClipboardEvent) => void;
  onCompositionend?: (payload: CompositionEvent) => void;
  onCompositionstart?: (payload: CompositionEvent) => void;
  onCompositionupdate?: (payload: CompositionEvent) => void;
  onDrag?: (payload: DragEvent) => void;
  onDragend?: (payload: DragEvent) => void;
  onDragenter?: (payload: DragEvent) => void;
  onDragexit?: (payload: DragEvent) => void;
  onDragleave?: (payload: DragEvent) => void;
  onDragover?: (payload: DragEvent) => void;
  onDragstart?: (payload: DragEvent) => void;
  onDrop?: (payload: DragEvent) => void;
  onFocus?: (payload: FocusEvent) => void;
  onFocusin?: (payload: FocusEvent) => void;
  onFocusout?: (payload: FocusEvent) => void;
  onBlur?: (payload: FocusEvent) => void;
  onChange?: (payload: Event) => void;
  onBeforeinput?: (payload: Event) => void;
  onInput?: (payload: Event) => void;
  onReset?: (payload: Event) => void;
  onSubmit?: (payload: Event) => void;
  onInvalid?: (payload: Event) => void;
  onLoad?: (payload: Event) => void;
  onError?: (payload: Event) => void;
  onKeydown?: (payload: KeyboardEvent) => void;
  onKeypress?: (payload: KeyboardEvent) => void;
  onKeyup?: (payload: KeyboardEvent) => void;
  onAuxclick?: (payload: MouseEvent) => void;
  onClick?: (payload: MouseEvent) => void;
  onContextmenu?: (payload: MouseEvent) => void;
  onDblclick?: (payload: MouseEvent) => void;
  onMousedown?: (payload: MouseEvent) => void;
  onMouseenter?: (payload: MouseEvent) => void;
  onMouseleave?: (payload: MouseEvent) => void;
  onMousemove?: (payload: MouseEvent) => void;
  onMouseout?: (payload: MouseEvent) => void;
  onMouseover?: (payload: MouseEvent) => void;
  onMouseup?: (payload: MouseEvent) => void;
  onAbort?: (payload: Event) => void;
  onCanplay?: (payload: Event) => void;
  onCanplaythrough?: (payload: Event) => void;
  onDurationchange?: (payload: Event) => void;
  onEmptied?: (payload: Event) => void;
  onEncrypted?: (payload: Event) => void;
  onEnded?: (payload: Event) => void;
  onLoadeddata?: (payload: Event) => void;
  onLoadedmetadata?: (payload: Event) => void;
  onLoadstart?: (payload: Event) => void;
  onPause?: (payload: Event) => void;
  onPlay?: (payload: Event) => void;
  onPlaying?: (payload: Event) => void;
  onProgress?: (payload: Event) => void;
  onRatechange?: (payload: Event) => void;
  onSeeked?: (payload: Event) => void;
  onSeeking?: (payload: Event) => void;
  onStalled?: (payload: Event) => void;
  onSuspend?: (payload: Event) => void;
  onTimeupdate?: (payload: Event) => void;
  onVolumechange?: (payload: Event) => void;
  onWaiting?: (payload: Event) => void;
  onSelect?: (payload: Event) => void;
  onScroll?: (payload: Event) => void;
  onScrollend?: (payload: Event) => void;
  onTouchcancel?: (payload: TouchEvent) => void;
  onTouchend?: (payload: TouchEvent) => void;
  onTouchmove?: (payload: TouchEvent) => void;
  onTouchstart?: (payload: TouchEvent) => void;
  onPointerdown?: (payload: PointerEvent) => void;
  onPointermove?: (payload: PointerEvent) => void;
  onPointerup?: (payload: PointerEvent) => void;
  onPointercancel?: (payload: PointerEvent) => void;
  onPointerenter?: (payload: PointerEvent) => void;
  onPointerleave?: (payload: PointerEvent) => void;
  onPointerover?: (payload: PointerEvent) => void;
  onPointerout?: (payload: PointerEvent) => void;
  onWheel?: (payload: WheelEvent) => void;
  onAnimationstart?: (payload: AnimationEvent) => void;
  onAnimationend?: (payload: AnimationEvent) => void;
  onAnimationiteration?: (payload: AnimationEvent) => void;
  onTransitionend?: (payload: TransitionEvent) => void;
  onTransitionstart?: (payload: TransitionEvent) => void;
  key?: PropertyKey;
  ref?: vue0.VNodeRef;
  ref_for?: boolean;
  ref_key?: string;
  onVnodeBeforeMount?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeMounted?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeBeforeUpdate?: ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void) | ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void)[];
  onVnodeUpdated?: ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void) | ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void)[];
  onVnodeBeforeUnmount?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeUnmounted?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
}, {}, {}, {}, {}, vue0.ComponentOptionsMixin, vue0.ComponentOptionsMixin, {}, string, vue0.PublicProps, Readonly<{
  foo: string;
  bar?: string;
  innerHTML?: string;
  class?: any;
  style?: vue0.StyleValue;
  accesskey?: string;
  contenteditable?: (boolean | "true" | "false") | "inherit" | "plaintext-only";
  contextmenu?: string;
  dir?: string;
  draggable?: boolean | "true" | "false";
  hidden?: (boolean | "true" | "false") | "" | "hidden" | "until-found";
  id?: string;
  inert?: boolean | "true" | "false";
  lang?: string;
  placeholder?: string;
  spellcheck?: boolean | "true" | "false";
  tabindex?: string | number;
  title?: string;
  translate?: "yes" | "no";
  radiogroup?: string;
  role?: string;
  about?: string;
  datatype?: string;
  inlist?: any;
  prefix?: string;
  property?: string;
  resource?: string;
  typeof?: string;
  vocab?: string;
  autocapitalize?: string;
  autocorrect?: string;
  autosave?: string;
  color?: string;
  itemprop?: string;
  itemscope?: boolean | "true" | "false";
  itemtype?: string;
  itemid?: string;
  itemref?: string;
  results?: string | number;
  security?: string;
  unselectable?: "on" | "off";
  inputmode?: "none" | "text" | "tel" | "url" | "email" | "numeric" | "decimal" | "search";
  is?: string;
  'aria-activedescendant'?: string;
  'aria-atomic'?: boolean | "true" | "false";
  'aria-autocomplete'?: "none" | "inline" | "list" | "both";
  'aria-busy'?: boolean | "true" | "false";
  'aria-checked'?: (boolean | "true" | "false") | "mixed";
  'aria-colcount'?: string | number;
  'aria-colindex'?: string | number;
  'aria-colspan'?: string | number;
  'aria-controls'?: string;
  'aria-current'?: (boolean | "true" | "false") | "page" | "step" | "location" | "date" | "time";
  'aria-describedby'?: string;
  'aria-details'?: string;
  'aria-disabled'?: boolean | "true" | "false";
  'aria-dropeffect'?: "none" | "copy" | "execute" | "link" | "move" | "popup";
  'aria-errormessage'?: string;
  'aria-expanded'?: boolean | "true" | "false";
  'aria-flowto'?: string;
  'aria-grabbed'?: boolean | "true" | "false";
  'aria-haspopup'?: (boolean | "true" | "false") | "menu" | "listbox" | "tree" | "grid" | "dialog";
  'aria-hidden'?: boolean | "true" | "false";
  'aria-invalid'?: (boolean | "true" | "false") | "grammar" | "spelling";
  'aria-keyshortcuts'?: string;
  'aria-label'?: string;
  'aria-labelledby'?: string;
  'aria-level'?: string | number;
  'aria-live'?: "off" | "assertive" | "polite";
  'aria-modal'?: boolean | "true" | "false";
  'aria-multiline'?: boolean | "true" | "false";
  'aria-multiselectable'?: boolean | "true" | "false";
  'aria-orientation'?: "horizontal" | "vertical";
  'aria-owns'?: string;
  'aria-placeholder'?: string;
  'aria-posinset'?: string | number;
  'aria-pressed'?: (boolean | "true" | "false") | "mixed";
  'aria-readonly'?: boolean | "true" | "false";
  'aria-relevant'?: "additions" | "additions removals" | "additions text" | "all" | "removals" | "removals additions" | "removals text" | "text" | "text additions" | "text removals";
  'aria-required'?: boolean | "true" | "false";
  'aria-roledescription'?: string;
  'aria-rowcount'?: string | number;
  'aria-rowindex'?: string | number;
  'aria-rowspan'?: string | number;
  'aria-selected'?: boolean | "true" | "false";
  'aria-setsize'?: string | number;
  'aria-sort'?: "none" | "ascending" | "descending" | "other";
  'aria-valuemax'?: string | number;
  'aria-valuemin'?: string | number;
  'aria-valuenow'?: string | number;
  'aria-valuetext'?: string;
  onCopy?: (payload: ClipboardEvent) => void;
  onCut?: (payload: ClipboardEvent) => void;
  onPaste?: (payload: ClipboardEvent) => void;
  onCompositionend?: (payload: CompositionEvent) => void;
  onCompositionstart?: (payload: CompositionEvent) => void;
  onCompositionupdate?: (payload: CompositionEvent) => void;
  onDrag?: (payload: DragEvent) => void;
  onDragend?: (payload: DragEvent) => void;
  onDragenter?: (payload: DragEvent) => void;
  onDragexit?: (payload: DragEvent) => void;
  onDragleave?: (payload: DragEvent) => void;
  onDragover?: (payload: DragEvent) => void;
  onDragstart?: (payload: DragEvent) => void;
  onDrop?: (payload: DragEvent) => void;
  onFocus?: (payload: FocusEvent) => void;
  onFocusin?: (payload: FocusEvent) => void;
  onFocusout?: (payload: FocusEvent) => void;
  onBlur?: (payload: FocusEvent) => void;
  onChange?: (payload: Event) => void;
  onBeforeinput?: (payload: Event) => void;
  onInput?: (payload: Event) => void;
  onReset?: (payload: Event) => void;
  onSubmit?: (payload: Event) => void;
  onInvalid?: (payload: Event) => void;
  onLoad?: (payload: Event) => void;
  onError?: (payload: Event) => void;
  onKeydown?: (payload: KeyboardEvent) => void;
  onKeypress?: (payload: KeyboardEvent) => void;
  onKeyup?: (payload: KeyboardEvent) => void;
  onAuxclick?: (payload: MouseEvent) => void;
  onClick?: (payload: MouseEvent) => void;
  onContextmenu?: (payload: MouseEvent) => void;
  onDblclick?: (payload: MouseEvent) => void;
  onMousedown?: (payload: MouseEvent) => void;
  onMouseenter?: (payload: MouseEvent) => void;
  onMouseleave?: (payload: MouseEvent) => void;
  onMousemove?: (payload: MouseEvent) => void;
  onMouseout?: (payload: MouseEvent) => void;
  onMouseover?: (payload: MouseEvent) => void;
  onMouseup?: (payload: MouseEvent) => void;
  onAbort?: (payload: Event) => void;
  onCanplay?: (payload: Event) => void;
  onCanplaythrough?: (payload: Event) => void;
  onDurationchange?: (payload: Event) => void;
  onEmptied?: (payload: Event) => void;
  onEncrypted?: (payload: Event) => void;
  onEnded?: (payload: Event) => void;
  onLoadeddata?: (payload: Event) => void;
  onLoadedmetadata?: (payload: Event) => void;
  onLoadstart?: (payload: Event) => void;
  onPause?: (payload: Event) => void;
  onPlay?: (payload: Event) => void;
  onPlaying?: (payload: Event) => void;
  onProgress?: (payload: Event) => void;
  onRatechange?: (payload: Event) => void;
  onSeeked?: (payload: Event) => void;
  onSeeking?: (payload: Event) => void;
  onStalled?: (payload: Event) => void;
  onSuspend?: (payload: Event) => void;
  onTimeupdate?: (payload: Event) => void;
  onVolumechange?: (payload: Event) => void;
  onWaiting?: (payload: Event) => void;
  onSelect?: (payload: Event) => void;
  onScroll?: (payload: Event) => void;
  onScrollend?: (payload: Event) => void;
  onTouchcancel?: (payload: TouchEvent) => void;
  onTouchend?: (payload: TouchEvent) => void;
  onTouchmove?: (payload: TouchEvent) => void;
  onTouchstart?: (payload: TouchEvent) => void;
  onPointerdown?: (payload: PointerEvent) => void;
  onPointermove?: (payload: PointerEvent) => void;
  onPointerup?: (payload: PointerEvent) => void;
  onPointercancel?: (payload: PointerEvent) => void;
  onPointerenter?: (payload: PointerEvent) => void;
  onPointerleave?: (payload: PointerEvent) => void;
  onPointerover?: (payload: PointerEvent) => void;
  onPointerout?: (payload: PointerEvent) => void;
  onWheel?: (payload: WheelEvent) => void;
  onAnimationstart?: (payload: AnimationEvent) => void;
  onAnimationend?: (payload: AnimationEvent) => void;
  onAnimationiteration?: (payload: AnimationEvent) => void;
  onTransitionend?: (payload: TransitionEvent) => void;
  onTransitionstart?: (payload: TransitionEvent) => void;
  key?: PropertyKey;
  ref?: vue0.VNodeRef;
  ref_for?: boolean;
  ref_key?: string;
  onVnodeBeforeMount?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeMounted?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeBeforeUpdate?: ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void) | ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void)[];
  onVnodeUpdated?: ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void) | ((vnode: vue0.VNode, oldVNode: vue0.VNode) => void)[];
  onVnodeBeforeUnmount?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
  onVnodeUnmounted?: ((vnode: vue0.VNode) => void) | ((vnode: vue0.VNode) => void)[];
}> & Readonly<{}>, {}, {}, {}, {}, string, vue0.ComponentProvideOptions, false, {}, any>;
//#endregion
export { _default as App };"
`;
